src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java

```diff
@@ -420,37 +429,60 @@ private String  subtitleForMediaType(int mediaType) {
 	private void runQuery(List<MediaBrowser.MediaItem> populateMe, int mediaType, MediaAdapter adapter) {
 		populateMe.clear();
 		try {
-			Cursor cursor = adapter.query();
-			Context context = getApplicationContext();
-			ContentResolver resolver = context.getContentResolver();
-
-			if (cursor == null) {
-				return;
+			final Context context = getApplicationContext();
+			final ContentResolver resolver = context.getContentResolver();
+			Cursor cursor;
+			if(mediaType == MediaUtils.TYPE_SONG) {
+				cursor = adapter.buildSongQuery(Song.FILLED_PROJECTION).runQuery(resolver);
+			} else {
+				cursor = adapter.query();
 			}
-
+			Song song = new Song(-1);
 			final int flags = (mediaType == MediaUtils.TYPE_SONG || mediaType == MediaUtils.TYPE_PLAYLIST) ? MediaBrowser.MediaItem.FLAG_PLAYABLE : MediaBrowser.MediaItem.FLAG_BROWSABLE;
 			final int count = cursor.getCount();
-			for (int j = 0; j != count; ++j) {
-				cursor.moveToPosition(j);
-				final String id = cursor.getString(0);
-				final String label = cursor.getString(2);
-				long mediaId = Long.parseLong(id);
-
-				Song song = MediaUtils.getSongByTypeId(resolver, mediaType, mediaId);
+			cursor.moveToFirst();
+			for (int j = 0; j < count; ++j) {
+				final Long id = cursor.getLong(0);
+				String title = cursor.getString(2);
+				Bitmap cover = null;
+				if(mediaType == MediaUtils.TYPE_SONG) {
+					if(j < 50) {
+						song.populate(cursor);
+						song.flags = 0;
+						if(song.isFilled()) {
```

Comment:
Ternary operation would save you 5 lines: `cover = song.isFilled() ? song.getSmallCover() : sFallbackBitmap`